src/Psy/ExecutionClosure.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+/*
+ * This file is part of Psy Shell.
+ *
+ * (c) 2012-2017 Justin Hileman
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Psy;
+
+/**
+ * The Psy Shell's execution scope.
+ */
+class ExecutionClosure
+{
+    const NOOP_INPUT = 'return null;';
+
+    private $closure;
+
+    /**
+     * @param Shell &$__psysh__
+     */
+    public function __construct(Shell &$__psysh__)
+    {
+        $exec = function () use (&$__psysh__) {
+            try {
+                // Restore execution scope variables
+                extract($__psysh__->getScopeVariables(false));
+
+                // evaluate the current code buffer
+                ob_start([$__psysh__, 'writeStdout'], 1);
+
+                set_error_handler([$__psysh__, 'handleError']);
+                $_ = eval($__psysh__->onExecute($__psysh__->flushCode() ?: ExecutionClosure::NOOP_INPUT));
+                restore_error_handler();
+
+                ob_end_flush();
+
+                $__psysh__->setScopeVariables(get_defined_vars());
+            } catch (\Throwable $_e) {
+                restore_error_handler();
+                if (ob_get_level() > 0) {
+                    ob_end_clean();
+                }
+
+                throw $_e;
+            } catch (\Exception $_e) {
+                restore_error_handler();
+                if (ob_get_level() > 0) {
+                    ob_end_clean();
+                }
+
+                throw $_e;
+            }
+
+            return $_;
+        };
+
+        if (self::bindClosure()) {
```

Comment:
Confusing method name perhaps? Better to name it `shouldBindClosure`?

Reply:
:+1: